Search through lectures

MatthiasReumann opened the pull request:
- Add search button on course page
- Add route /search?q=<query string>
- Initally search for titles, sections, descriptions 


[OpenSource Lab]

MatthiasReumann committed:
Initial commit

joschahenningsen commented:
https://mariadb.com/kb/en/full-text-index-overview/ might me helpful here. https://nav.tum.sexy uses Meilisearch as search engine. Worth a look but imo that's overkill here, even when searching through chats or auto-generated transcripts.

MatthiasReumann committed:
Merge branch 'dev' into enhancement/search-through-lectures

MatthiasReumann committed:
Add simple search api

MatthiasReumann committed:
Merge branch 'dev' into enhancement/search-through-lectures

MatthiasReumann committed:
Add streams search

MatthiasReumann committed:
Add search bar dummy

MatthiasReumann committed:
Add courseid to query

MatthiasReumann committed:
Merge branch 'dev' into enhancement/search-through-lectures

# Conflicts:
#	api/router.go

MatthiasReumann committed:
Update search api and search ui

MatthiasReumann commented:
**Update/Documentation:** 
I added a new API Endpoint `/api/search/streams/<searchDomain>` with `<searchDomain>` := `sections` | `streams` | `chats` so that in the future one could easily add another search endpoint like `/api/search/courses` and don't have to worry about bad naming. 

Each streams-search Endpoint returns a JSON encoded object of 
```
type searchResponse struct {
	StreamIds []uint
}
```

I've utilised the full-text-search capabilities of MariaDB, which meant indices to columns were added. 

Frontend-wise, the client makes calls the respective endpoints, merges the results and makes a call to `/api/streams/:streamID` and displays that information. I've updated the `/api/streams/:streamID`-Endpoint insofar that non-admins can also call `GET` to retrieve a reduced version of the stream JSON object:

```
response =
  gin.H{"name": stream.Name,
  "description":  stream.Description,
  "friendlyTime": stream.FriendlyTime()}
```

**UI:**
<img width="631" alt="Screenshot 2022-05-15 at 10 51 37" src="https://user-images.githubusercontent.com/29633518/168464881-702f435f-4a82-4909-b613-e3b42599a425.png">
<img width="627" alt="Screenshot 2022-05-15 at 10 53 00" src="https://user-images.githubusercontent.com/29633518/168464885-8cdbd958-0a77-4758-af7e-fe998ab2ae32.png">


**Follow-up:**
Once I've done enough research for voice-to-text, I will start a another PR which addresses the search in that regard.

joschahenningsen commented on model/stream.go:
```diff
@@ -24,8 +24,8 @@ const (
-	Name             string
-	Description      string
+	Name             string `gorm:"index:,class:FULLTEXT"`
+	Description      string `gorm:"type:text;index:,class:FULLTEXT"`
```
Just wondering: Why does this need `type:text`?

MatthiasReumann replied:
`FULLTEXT` does only work on `CHAR`, `VARCHAR`, AND `TEXT`

CommanderStorm commented on web/ts/stream-search.ts:
```diff
@@ -0,0 +1,77 @@
+                    this.loading = false;
+                    this.results = this.getResults(searchResponses.map((sr) => sr.StreamIds));
+                });
+            }, 250);
```
Is the Timeout here really intended?
> The `setTimeout()` method calls a function after a number of milliseconds

so as far as I read this, there is an delay of 250ms between the user inputting something and the search being executed?

MatthiasReumann replied:
The Timeout is there so that the function isn't called after every `keyup`, but rather the last. 
So for example if I type `h-e-l-l-o`:
After `h`,`e`,`l`,`l` `clearTimeout(timout) `resets the timeout, and after the letter `o` it executes the callback of `setTimeout`.

joschahenningsen replied:
How expensive is the search request? I don't think it would be a problem if this was fetched on every keystroke.  But please make sure, that the request is only sent if the input has changed, not on keys like `esc`, `shift`, `tab` etc.

CommanderStorm commented on web/ts/stream-search.ts:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private async search(q: string, searchDomain: SearchDomain): Promise<searchResponse> {
+        return await fetch(`/api/search/streams/${searchDomain}?q=${q}&courseId=${this.courseId}`)
```
you might want to use something like a cached fetch (see https://github.com/TUM-Dev/navigatum/blob/db0a6e7c0662c1e6211052eb18ebdefefbf3d0c1/webclient/src/core.js#L8 for inspiration) to improve the search experience. 
In a context where responsiveness matters, doing unnessesary work seems wrong.

CommanderStorm commented on web/template/course-overview.gohtml:
```diff
@@ -6,10 +6,50 @@
+                <i x-cloak x-show="streamSearch.loading"
+                   class="fas fa-spinner animate-spin text-lg text-gray-500 dark:text-gray-600"></i>
+            </div>
+            <input @keyup="(e) => streamSearch.performSearch(e.target.value)"
```
You may want to include some extra keys on keydown
See https://github.com/TUM-Dev/navigatum/blob/db0a6e7c0662c1e6211052eb18ebdefefbf3d0c1/webclient/src/modules/autocomplete.js#L95

CommanderStorm commented on web/template/course-overview.gohtml:
```diff
@@ -6,10 +6,50 @@
+            </div>
+            <input @keyup="(e) => streamSearch.performSearch(e.target.value)"
+                   name="lecture-search-input"
+                   placeholder="Search for lectures"
```
Kind of personal preference
```suggestion
                   placeholder="Search for a lecture"
```

CommanderStorm commented on web/ts/stream-search.ts:
```diff
@@ -0,0 +1,77 @@
+            this.loading = true;
+            this.timeout = setTimeout(() => {
+                Promise.all([
+                    this.search(q, SearchDomain.VideoSections),
+                    this.search(q, SearchDomain.Streams),
+                    this.search(q, SearchDomain.Chats),
```
I find it kind of strange to impose a fixed order of search results.
For NavigaTUM we did a lot of work on this, while for this code this seems to be hardcoded. This may mean that less relevant search results come first..

I can however see that this is a good shortcut, for implementing this, if this feature is not meant to be something core of the platform.
I would classify this feature as a "nice to have", thus this shortcut is imo okay.

This could also be because of the quality/speed tradeoff of the fulltext search never returns less relevant search results. Or the added complexity is would not really be worth the added tech dept.
In these cases a sourcecode comment documenting this would be appreciated xD

MatthiasReumann replied:
I generally think that the (limited) full-text-search of MariaDB is only an intermediate solution and something like ElasticSearch, Meilisearch, or similar is needed if, as you say, search becomes an integral part of TUMLive. 

However, I could add the `relevance` (a float value) to the `searchResponse` object, which would deal with the problem of your first paragraph.

joschahenningsen replied:
relevance-wise I think lecture-name > description > chat. Weighing this with some factor and the relevance from mariadb could work out quite well. If we really want to create a real search engine for TUM-Live, we could have a look at https://blevesearch.com/ (But in a follow up :D).

Also I agree with Frank, we should have this be one request instead of three. Maybe by adding the searchDomains to the URL like this: `?searchDomains=a,b,c` or by adding a `all` search domain.

CommanderStorm commented on web/ts/stream-search.ts:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    getResults(results: number[][]): object[] {
+        const all = []; // create key value pair from array of array
```
This should be a better documentation of this function
```suggestion 
        // flatten array of array, preserving only the first occurrence of each streamId
        const all = [];
```

CommanderStorm reviewed (commented):
Note that these are just some comments from a fellow radom internet stranger (read: I am not a maintainer of this project).
- I am not qualified to talk about Go, so I am not Go-ing to do that
- nor am I particularly qualified to talk about Typescript
- I have not tested your code (I just had a look), so this is not an attestation, that this works

I have left some inline comments.

About the engenering:
- I find it a bit strange to not directly fetch the data you would like to deliver in the search results in the search endpoint, but this may be a good idea, depending on how much you see this endpoint being used.
- collecting multiple requests in JS seems also kind of strange imo

joschahenningsen commented on api/stream.go:
```diff
@@ -200,17 +200,33 @@ func (r streamRoutes) getStream(c *gin.Context) {
+	}
+
+	var response gin.H
+	if tumLiveContext.User != nil && tumLiveContext.UserIsAdmin() {
```
The "larger" response should be sent to lecturers of the course as well, not only server admins. Also this needs a comment or two, which different purposes the if and else branches serve. 
But ideally, as stated below we don't need this endpoint.

joschahenningsen commented on dao/chat.go:
```diff
@@ -193,3 +196,18 @@ func (d chatDao) ToggleLike(userID uint, chatID uint) error {
 	return DB.Table("polls").Where("stream_id = ? AND active", streamID).Update("active", false).Error
 }
+
+func (d chatDao) Search(q string, courseId uint) ([]uint, error) {
```
Can you add a doc- comment and/or name the return params?

joschahenningsen commented on dao/streams.go:
```diff
@@ -415,3 +416,17 @@ func (d streamsDao) DeleteStreamsWithTumID(ids []uint) {
 	})
 }
+
+func (d streamsDao) Search(q string, courseId uint) ([]uint, error) {
```
Can you add a doc- comment and/or name the return params?

joschahenningsen commented on dao/video-section.go:
```diff
@@ -37,3 +40,18 @@ func (d videoSectionDao) GetByStreamId(streamID uint) ([]model.VideoSection, err
 	return sections, err
 }
+
+func (d videoSectionDao) Search(q string, courseId uint) ([]uint, error) {
```
Here as well: Can you add a doc- comment and/or name the return params?

joschahenningsen reviewed (changes requested):
Looks quite good to me, just some comments. Also thanks @CommanderStorm for the input :blush:

MatthiasReumann committed:
Minor changes

MatthiasReumann commented:
> 

Is `VL1` and `VL2` a VoD? Since, currently,  it only displays VoD results.

MatthiasReumann committed:
Update search

MatthiasReumann committed:
Remove unused functions

joschahenningsen commented:
> Is VL1 and VL2 a VoD? Since, currently, it only displays VoD results.

yes, I'm testing on the database referenced in the readme and applied no changes to it except the automigration from this pr

MatthiasReumann committed:
Merge branch 'dev' into enhancement/search-through-lectures

# Conflicts:
#	api/stream.go
#	web/template/course-overview.gohtml

MatthiasReumann committed:
Add cached fetch

MatthiasReumann commented:
**Update:** 

First all thanks to both of you for the great feedback ❤️ @CommanderStorm @joschahenningsen  

- Added SearchDAO: One query for everything (Endpoint: `/api/search/streams`), no domains anymore, sorted by relevance, no `GET` requests on frontend side.
- Added Cached Fetch
- Added dark mode styling 
- Added Keydown & Keyup support

MatthiasReumann committed:
Merge branch 'dev' into enhancement/search-through-lectures

joschahenningsen reviewed (approved):
Sorry for the long wait, this looks good! Can you just restore the run configuration?

MatthiasReumann committed:
Add run config

MatthiasReumann committed:
Format xml

joschahenningsen commented:
Thanks! :)
